File: scripts/chiusura-konto-1099.py
#!/usr/bin/env python3
"""
AUTOMAZIONE CHIUSURA KONTO 1099 TRANSFERKONTO

Chiude il conto 1099 (CHF -60,842.41) su Patrimonio Netto
Movimenti del 31.01.2024 - Correzioni post-migrazione 2023

@author Process Automator
@date 2025-11-15
"""

import odoorpc
from datetime import date

# ============================================================================
# CONFIGURAZIONE
# ============================================================================

ODOO_HOST = 'lapadevadmin-lapa-v2-staging-2406-25408900.dev.odoo.com'
ODOO_DB = 'lapadevadmin-lapa-v2-staging-2406-25408900'
ODOO_USER = 'paul@lapa.ch'
ODOO_PASS = 'lapa201180'

# ============================================================================
# FUNZIONI
# ============================================================================

def print_separator(char='=', length=60):
    """Stampa separatore"""
    print(char * length)

def print_step(step_num, title):
    """Stampa titolo step"""
    print(f"\n📊 STEP {step_num}: {title}\n")
    print_separator()

# ============================================================================
# MAIN
# ============================================================================

def main():
    print_separator()
    print("  AUTOMAZIONE CHIUSURA KONTO 1099 TRANSFERKONTO")
    print_separator()

    # ========================================================================
    # CONNESSIONE
    # ========================================================================

    print("\n🔐 Connessione a Odoo...")
    try:
        odoo = odoorpc.ODOO(ODOO_HOST, protocol='jsonrpc+ssl', port=443)
        odoo.login(ODOO_DB, ODOO_USER, ODOO_PASS)
        print(f"✅ Connesso! UID: {odoo.env.uid}")
    except Exception as e:
        print(f"❌ Errore connessione: {e}")
        return 1

    # ========================================================================
    # STEP 1: ANALISI KONTO 1099
    # ========================================================================

    print_step(1, "Analisi Konto 1099 Transferkonto")

    Account = odoo.env['account.account']
    MoveLine = odoo.env['account.move.line']

    # Cerca conto 1099
    account_ids = Account.search([('code', '=', '1099')])

    if not account_ids:
        print("❌ Conto 1099 non trovato!")
        return 1

    konto1099 = Account.browse(account_ids[0])

    print(f"\n✅ Conto trovato:")
    print(f"   ID: {konto1099.id}")
    print(f"   Codice: {konto1099.code}")
    print(f"   Nome: {konto1099.name}")
    print(f"   Tipo: {konto1099.account_type}")

    # Cerca movimenti
    line_ids = MoveLine.search([('account_id', '=', konto1099.id)])
    all_lines = MoveLine.browse(line_ids)

    print(f"\n📋 Movimenti totali: {len(all_lines)}")

    # Filtra quelli del 31.01.2024
    lines_jan = [line for line in all_lines if str(line.date) == '2024-01-31']
    print(f"   di cui del 31.01.2024: {len(lines_jan)}")
    print('-' * 60)

    total_debit = 0
    total_credit = 0

    for idx, line in enumerate(lines_jan, 1):
        print(f"\n{idx}. Move Line ID: {line.id}")
        print(f"   Data: {line.date}")
        print(f"   Descrizione: {line.name or 'N/A'}")
        print(f"   Dare: CHF {line.debit:.2f}")
        print(f"   Avere: CHF {line.credit:.2f}")
        print(f"   Saldo: CHF {line.balance:.2f}")

        total_debit += line.debit
        total_credit += line.credit

    net_balance = total_debit - total_credit
    current_balance = konto1099.current_balance

    print(f"\n{'=' * 60}")
    print(f"📊 RIEPILOGO (31.01.2024):")
    print(f"   Totale Dare: CHF {total_debit:.2f}")
    print(f"   Totale Avere: CHF {total_credit:.2f}")
    print(f"   Saldo Netto: CHF {net_balance:.2f}")
    print(f"\n💡 Saldo attuale conto: CHF {current_balance:.2f}")
    print('=' * 60)

    # ========================================================================
    # STEP 2: TROVA CONTO PATRIMONIO NETTO
    # ========================================================================

    print_step(2, "Identifica Conto Patrimonio Netto")

    # Cerca conti equity
    equity_ids = Account.search([('account_type', 'in', ['equity', 'equity_unaffected'])])
    equity_accounts = Account.browse(equity_ids)

    print(f"\n📋 Conti Equity disponibili: {len(equity_accounts)}")

    if not equity_accounts:
        print("❌ Nessun conto Equity trovato!")
        return 1

    # Mostra i primi 10
    for idx, acc in enumerate(equity_accounts[:10], 1):
        print(f"   {idx}. [{acc.code}] {acc.name}")

    # Cerca specificamente conti di apertura/differenze
    target_account = None

    for acc in equity_accounts:
        if acc.code in ['2979', '2980']:
            target_account = acc
            break

    if not target_account:
        for acc in equity_accounts:
            name_lower = acc.name.lower()
            if any(keyword in name_lower for keyword in ['eröffnung', 'differenz', 'apertura', 'opening']):
                target_account = acc
                break

    if not target_account:
        for acc in equity_accounts:
            if acc.account_type == 'equity_unaffected':
                target_account = acc
                break

    if not target_account:
        target_account = equity_accounts[0]

    print(f"\n✅ Conto selezionato:")
    print(f"   ID: {target_account.id}")
    print(f"   Codice: {target_account.code}")
    print(f"   Nome: {target_account.name}")
    print(f"   Tipo: {target_account.account_type}")

    # ========================================================================
    # STEP 3: CREA REGISTRAZIONE
    # ========================================================================

    print_step(3, "Creazione Registrazione di Chiusura")

    Journal = odoo.env['account.journal']

    # Cerca journal generale
    journal_ids = Journal.search([('type', '=', 'general')], limit=1)

    if not journal_ids:
        print("❌ Nessun journal generale trovato!")
        return 1

    journal = Journal.browse(journal_ids[0])
    print(f"\n📘 Journal: {journal.code} - {journal.name}")

    importo = abs(current_balance)

    print(f"\n💡 Saldo da chiudere: CHF {current_balance:.2f}")
    print(f"   ({'Credito' if current_balance < 0 else 'Debito'})")

    # Prepara righe
    if current_balance < 0:
        # Saldo credito → Dare 1099, Avere Equity
        line1099 = {
            'account_id': konto1099.id,
            'name': 'Chiusura Transferkonto su Patrimonio Netto',
            'debit': importo,
            'credit': 0
        }
        line_equity = {
            'account_id': target_account.id,
            'name': 'Chiusura Transferkonto da conto 1099',
            'debit': 0,
            'credit': importo
        }
    else:
        # Saldo debito → Avere 1099, Dare Equity
        line1099 = {
            'account_id': konto1099.id,
            'name': 'Chiusura Transferkonto su Patrimonio Netto',
            'debit': 0,
            'credit': importo
        }
        line_equity = {
            'account_id': target_account.id,
            'name': 'Chiusura Transferkonto da conto 1099',
            'debit': importo,
            'credit': 0
        }

    print(f"\n📋 Registrazione da creare:")
    print(f"   Data: {date.today()}")
    print(f"   Importo: CHF {importo:.2f}")
    print(f"\n   Riga 1: {konto1099.code} - {konto1099.name}")
    print(f"      D: CHF {line1099['debit']:.2f} | A: CHF {line1099['credit']:.2f}")
    print(f"\n   Riga 2: {target_account.code} - {target_account.name}")
    print(f"      D: CHF {line_equity['debit']:.2f} | A: CHF {line_equity['credit']:.2f}")

    print("\n⏳ Creazione registrazione in Odoo...")

    Move = odoo.env['account.move']

    move_data = {
        'journal_id': journal.id,
        'date': str(date.today()),
        'ref': 'Chiusura Konto 1099 Transferkonto - Correzioni post-migrazione 2023',
        'line_ids': [
            (0, 0, line1099),
            (0, 0, line_equity)
        ]
    }

    try:
        move_id = Move.create(move_data)
        print(f"✅ Registrazione creata: ID {move_id}")

        move = Move.browse(move_id)

        print(f"\n📄 Dettagli:")
        print(f"   Numero: {move.name}")
        print(f"   Stato: {move.state}")
        print(f"   Data: {move.date}")
        print(f"   Riferimento: {move.ref}")

    except Exception as e:
        print(f"\n❌ Errore nella creazione: {e}")
        return 1

    # ========================================================================
    # STEP 4: VERIFICA
    # ========================================================================

    print_step(4, "Verifica Saldo Finale")

    # Ricarica il conto per avere il saldo aggiornato
    konto1099 = Account.browse(konto1099.id)
    saldo_finale = konto1099.current_balance

    print(f"\n📊 Conto: {konto1099.code} - {konto1099.name}")
    print(f"   Saldo Finale: CHF {saldo_finale:.2f}")

    if abs(saldo_finale) < 0.01:
        print("\n🎉 SUCCESSO! Saldo = 0.00")
        verifica_ok = True
    else:
        print("\n⚠️  Saldo diverso da 0.00")
        print("   Potrebbe essere necessario validare la registrazione.")
        verifica_ok = False

    # ========================================================================
    # RIEPILOGO
    # ========================================================================

    print(f"\n{'=' * 60}")
    print("  RIEPILOGO OPERAZIONE")
    print('=' * 60)
    print(f"\n✅ Conto chiuso: {konto1099.code} - {konto1099.name}")
    print(f"✅ Importo: CHF {importo:.2f}")
    print(f"✅ Su conto: {target_account.code} - {target_account.name}")
    print(f"✅ Registrazione: {move.name} ({move.state})")
    print(f"{'✅' if verifica_ok else '⚠️ '} Saldo finale: {'CHF 0.00' if verifica_ok else 'Da verificare'}")

    if move.state == 'draft':
        print(f"\n⚠️  IMPORTANTE: La registrazione è in stato DRAFT")
        print(f"   Vai su Odoo e valida la registrazione {move.name}")

    print(f"\n{'=' * 60}")
    print("  ✅ AUTOMAZIONE COMPLETATA")
    print('=' * 60)
    print()

    return 0

if __name__ == '__main__':
    try:
        exit(main())
    except Exception as e:
        print(f"\n❌ ERRORE: {e}")
        import traceback
        traceback.print_exc()
        exit(1)

File: scripts/chiusura-konto-1099.js
/**
 * AUTOMAZIONE CHIUSURA KONTO 1099 TRANSFERKONTO
 *
 * Chiude il conto 1099 (CHF -60,842.41) su Patrimonio Netto
 * Movimenti del 31.01.2024 - Correzioni post-migrazione 2023
 *
 * @author Process Automator
 * @date 2025-11-15
 */

const Odoo = require('odoo-xmlrpc');
const fs = require('fs');

// ============================================================================
// CONFIGURAZIONE ODOO
// ============================================================================

const odoo = new Odoo({
  url: 'lapadevadmin-lapa-v2-staging-2406-25408900.dev.odoo.com',
  port: 443,
  db: 'lapadevadmin-lapa-v2-staging-2406-25408900',
  username: 'paul@lapa.ch',
  password: 'lapa201180',
  secure: true  // Forza HTTPS
});

let uid = null;

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function connect() {
  return new Promise((resolve, reject) => {
    odoo.connect((err) => {
      if (err) reject(err);
      else {
        uid = odoo.uid;
        resolve(uid);
      }
    });
  });
}

function executeKw(model, method, params = [], kwargs = {}) {
  return new Promise((resolve, reject) => {
    odoo.execute_kw(model, method, params, kwargs, (err, value) => {
      if (err) reject(err);
      else resolve(value);
    });
  });
}

async function searchRead(model, domain, fields) {
  return await executeKw(model, 'search_read', [domain], { fields, limit: 1000 });
}

async function create(model, values) {
  return await executeKw(model, 'create', [values]);
}

async function read(model, ids, fields) {
  return await executeKw(model, 'read', [ids], { fields });
}

// ============================================================================
// STEP 1: ANALISI KONTO 1099
// ============================================================================

async function analizzaKonto1099() {
  console.log('\n📊 STEP 1: Analisi Konto 1099 Transferkonto\n');
  console.log('='.repeat(60));

  // Cerca il conto 1099
  const accounts = await searchRead(
    'account.account',
    [['code', '=', '1099']],
    ['id', 'name', 'code', 'account_type', 'current_balance']
  );

  if (accounts.length === 0) {
    throw new Error('❌ Conto 1099 non trovato!');
  }

  const konto1099 = accounts[0];
  console.log(`\n✅ Conto trovato:`);
  console.log(`   ID: ${konto1099.id}`);
  console.log(`   Codice: ${konto1099.code}`);
  console.log(`   Nome: ${konto1099.name}`);
  console.log(`   Tipo: ${konto1099.account_type}`);

  // Cerca tutti i movimenti
  const allMoves = await searchRead(
    'account.move.line',
    [['account_id', '=', konto1099.id]],
    ['id', 'date', 'name', 'debit', 'credit', 'balance', 'move_id', 'ref']
  );

  console.log(`\n📋 Movimenti totali: ${allMoves.length}`);

  // Filtra quelli del 31.01.2024
  const movesJan = allMoves.filter(m => m.date === '2024-01-31');
  console.log(`   di cui del 31.01.2024: ${movesJan.length}`);
  console.log('-'.repeat(60));

  let totalDebit = 0;
  let totalCredit = 0;

  movesJan.forEach((line, idx) => {
    console.log(`\n${idx + 1}. Move Line ID: ${line.id}`);
    console.log(`   Data: ${line.date}`);
    console.log(`   Descrizione: ${line.name || 'N/A'}`);
    console.log(`   Dare: CHF ${line.debit.toFixed(2)}`);
    console.log(`   Avere: CHF ${line.credit.toFixed(2)}`);
    console.log(`   Saldo: CHF ${line.balance.toFixed(2)}`);

    totalDebit += line.debit;
    totalCredit += line.credit;
  });

  const netBalance = totalDebit - totalCredit;
  const currentBalance = konto1099.current_balance || netBalance;

  console.log('\n' + '='.repeat(60));
  console.log(`📊 RIEPILOGO (31.01.2024):`);
  console.log(`   Totale Dare: CHF ${totalDebit.toFixed(2)}`);
  console.log(`   Totale Avere: CHF ${totalCredit.toFixed(2)}`);
  console.log(`   Saldo Netto: CHF ${netBalance.toFixed(2)}`);
  console.log(`\n💡 Saldo attuale conto: CHF ${currentBalance.toFixed(2)}`);
  console.log('='.repeat(60));

  return {
    account: konto1099,
    moves: movesJan,
    allMoves,
    totalDebit,
    totalCredit,
    netBalance,
    currentBalance
  };
}

// ============================================================================
// STEP 2: TROVA CONTO PATRIMONIO NETTO
// ============================================================================

async function trovaContoPatrimonioNetto() {
  console.log('\n💰 STEP 2: Identifica Conto Patrimonio Netto\n');
  console.log('='.repeat(60));

  // Cerca conti equity
  const equityAccounts = await searchRead(
    'account.account',
    [['account_type', 'in', ['equity', 'equity_unaffected']]],
    ['id', 'name', 'code', 'account_type']
  );

  console.log(`\n📋 Conti Equity disponibili: ${equityAccounts.length}`);

  if (equityAccounts.length === 0) {
    throw new Error('❌ Nessun conto Equity trovato!');
  }

  // Mostra i primi 10
  equityAccounts.slice(0, 10).forEach((acc, idx) => {
    console.log(`   ${idx + 1}. [${acc.code}] ${acc.name}`);
  });

  // Cerca specificamente conti di apertura/differenze
  let targetAccount = equityAccounts.find(a => a.code === '2979' || a.code === '2980');

  if (!targetAccount) {
    targetAccount = equityAccounts.find(a =>
      a.name.toLowerCase().includes('eröffnung') ||
      a.name.toLowerCase().includes('differenz') ||
      a.name.toLowerCase().includes('apertura') ||
      a.name.toLowerCase().includes('opening')
    );
  }

  if (!targetAccount) {
    targetAccount = equityAccounts.find(a => a.account_type === 'equity_unaffected');
  }

  if (!targetAccount) {
    targetAccount = equityAccounts[0];
  }

  console.log(`\n✅ Conto selezionato:`);
  console.log(`   ID: ${targetAccount.id}`);
  console.log(`   Codice: ${targetAccount.code}`);
  console.log(`   Nome: ${targetAccount.name}`);
  console.log(`   Tipo: ${targetAccount.account_type}`);

  return targetAccount;
}

// ============================================================================
// STEP 3: CREA REGISTRAZIONE
// ============================================================================

async function creaRegistrazioneChiusura(konto1099, kontoEquity, currentBalance) {
  console.log('\n📝 STEP 3: Creazione Registrazione di Chiusura\n');
  console.log('='.repeat(60));

  // Cerca journal generale
  const journals = await searchRead(
    'account.journal',
    [['type', '=', 'general']],
    ['id', 'name', 'code']
  );

  if (journals.length === 0) {
    throw new Error('❌ Nessun journal generale trovato!');
  }

  const journal = journals[0];
  console.log(`\n📘 Journal: ${journal.code} - ${journal.name}`);

  const importo = Math.abs(currentBalance);

  console.log(`\n💡 Saldo da chiudere: CHF ${currentBalance.toFixed(2)}`);
  console.log(`   (${currentBalance < 0 ? 'Credito' : 'Debito'})`);

  // Se saldo è negativo (credito), dobbiamo dare il conto per portarlo a 0
  let line1099, lineEquity;

  if (currentBalance < 0) {
    // Saldo credito → Dare 1099, Avere Equity
    line1099 = {
      account_id: konto1099.id,
      name: 'Chiusura Transferkonto su Patrimonio Netto',
      debit: importo,
      credit: 0
    };
    lineEquity = {
      account_id: kontoEquity.id,
      name: 'Chiusura Transferkonto da conto 1099',
      debit: 0,
      credit: importo
    };
  } else {
    // Saldo debito → Avere 1099, Dare Equity
    line1099 = {
      account_id: konto1099.id,
      name: 'Chiusura Transferkonto su Patrimonio Netto',
      debit: 0,
      credit: importo
    };
    lineEquity = {
      account_id: kontoEquity.id,
      name: 'Chiusura Transferkonto da conto 1099',
      debit: importo,
      credit: 0
    };
  }

  console.log(`\n📋 Registrazione da creare:`);
  console.log(`   Data: ${new Date().toISOString().split('T')[0]}`);
  console.log(`   Importo: CHF ${importo.toFixed(2)}`);
  console.log(`\n   Riga 1: ${konto1099.code} - ${konto1099.name}`);
  console.log(`      D: CHF ${line1099.debit.toFixed(2)} | A: CHF ${line1099.credit.toFixed(2)}`);
  console.log(`\n   Riga 2: ${kontoEquity.code} - ${kontoEquity.name}`);
  console.log(`      D: CHF ${lineEquity.debit.toFixed(2)} | A: CHF ${lineEquity.credit.toFixed(2)}`);

  const moveData = {
    journal_id: journal.id,
    date: new Date().toISOString().split('T')[0],
    ref: 'Chiusura Konto 1099 Transferkonto - Correzioni post-migrazione 2023',
    line_ids: [
      [0, 0, line1099],
      [0, 0, lineEquity]
    ]
  };

  console.log('\n⏳ Creazione registrazione in Odoo...');

  try {
    const moveId = await create('account.move', moveData);
    console.log(`✅ Registrazione creata: ID ${moveId}`);

    const move = await read('account.move', [moveId], ['name', 'state', 'date', 'ref']);

    console.log(`\n📄 Dettagli:`);
    console.log(`   Numero: ${move[0].name}`);
    console.log(`   Stato: ${move[0].state}`);
    console.log(`   Data: ${move[0].date}`);
    console.log(`   Riferimento: ${move[0].ref}`);

    return { moveId, move: move[0] };
  } catch (error) {
    console.error(`\n❌ Errore nella creazione: ${error.message}`);
    throw error;
  }
}

// ============================================================================
// STEP 4: VERIFICA
// ============================================================================

async function verificaSaldoFinale(accountId) {
  console.log('\n✅ STEP 4: Verifica Saldo Finale\n');
  console.log('='.repeat(60));

  const accounts = await read('account.account', [accountId], ['id', 'name', 'code', 'current_balance']);

  const account = accounts[0];
  const saldo = account.current_balance || 0;

  console.log(`\n📊 Conto: ${account.code} - ${account.name}`);
  console.log(`   Saldo Finale: CHF ${saldo.toFixed(2)}`);

  if (Math.abs(saldo) < 0.01) {
    console.log('\n🎉 SUCCESSO! Saldo = 0.00');
    return true;
  } else {
    console.log('\n⚠️  Saldo diverso da 0.00');
    console.log('   Potrebbe essere necessario validare la registrazione.');
    return false;
  }
}

// ============================================================================
// DOCUMENTAZIONE
// ============================================================================

function generaDoc(analisi, kontoEquity, registrazione, verificaOk) {
  const ODOO_URL = 'https://lapadevadmin-lapa-v2-staging-2406-25408900.dev.odoo.com';
  const ODOO_DB = 'lapadevadmin-lapa-v2-staging-2406-25408900';
  const ODOO_USERNAME = 'paul@lapa.ch';

  return `
${'='.repeat(80)}
CHIUSURA KONTO 1099 TRANSFERKONTO - DOCUMENTAZIONE
${'='.repeat(80)}

Data: ${new Date().toLocaleString('it-CH')}
Eseguito da: Process Automator

${'-'.repeat(80)}
1. SITUAZIONE INIZIALE
${'-'.repeat(80)}

Conto: ${analisi.account.code} - ${analisi.account.name}
Saldo: CHF ${analisi.currentBalance.toFixed(2)}

Movimenti analizzati del 31.01.2024: ${analisi.moves.length}
- Totale Dare: CHF ${analisi.totalDebit.toFixed(2)}
- Totale Avere: CHF ${analisi.totalCredit.toFixed(2)}

Origine: Correzioni post-migrazione software 2023

${'-'.repeat(80)}
2. CONTO DI DESTINAZIONE
${'-'.repeat(80)}

Conto Patrimonio Netto: ${kontoEquity.code} - ${kontoEquity.name}
Tipo: ${kontoEquity.account_type}

Motivazione: Come da istruzioni commercialista, il Transferkonto deve essere
             chiuso a 0.00 su un conto di Patrimonio Netto (Eigenkapital).

${'-'.repeat(80)}
3. REGISTRAZIONE CONTABILE CREATA
${'-'.repeat(80)}

Numero: ${registrazione.move.name}
Data: ${registrazione.move.date}
Stato: ${registrazione.move.state}
ID: ${registrazione.moveId}

Importo: CHF ${Math.abs(analisi.currentBalance).toFixed(2)}

Movimenti:
- ${analisi.currentBalance < 0 ? 'DARE' : 'AVERE'} ${analisi.account.code}: CHF ${Math.abs(analisi.currentBalance).toFixed(2)}
- ${analisi.currentBalance < 0 ? 'AVERE' : 'DARE'} ${kontoEquity.code}: CHF ${Math.abs(analisi.currentBalance).toFixed(2)}

${'-'.repeat(80)}
4. RISULTATO
${'-'.repeat(80)}

Saldo finale Konto 1099: ${verificaOk ? 'CHF 0.00 ✅' : 'Da verificare ⚠️'}

${registrazione.move.state === 'draft' ? `
⚠️  AZIONE RICHIESTA:
La registrazione ${registrazione.move.name} è in stato DRAFT.

Per completare:
1. Accedi a Odoo: ${ODOO_URL}
2. Vai a: Contabilità > Registrazioni Contabili
3. Cerca: ${registrazione.move.name}
4. Clicca "Validate" (Valida)
5. Verifica saldo conto 1099 = 0.00
` : '✅ Operazione completata'}

${'-'.repeat(80)}
5. RIFERIMENTI
${'-'.repeat(80)}

Odoo URL: ${ODOO_URL}
Database: ${ODOO_DB}
User: ${ODOO_USERNAME}

Conto 1099 ID: ${analisi.account.id}
Conto Equity ID: ${kontoEquity.id}
Registrazione ID: ${registrazione.moveId}

${'='.repeat(80)}
`;
}

// ============================================================================
// MAIN
// ============================================================================

async function main() {
  console.log('\n' + '='.repeat(60));
  console.log('  AUTOMAZIONE CHIUSURA KONTO 1099 TRANSFERKONTO');
  console.log('='.repeat(60));

  try {
    // Connessione
    console.log('\n🔐 Connessione a Odoo...');
    await connect();
    console.log(`✅ Connesso! UID: ${uid}`);

    // Esegui gli step
    const analisi = await analizzaKonto1099();
    const kontoEquity = await trovaContoPatrimonioNetto();
    const registrazione = await creaRegistrazioneChiusura(
      analisi.account,
      kontoEquity,
      analisi.currentBalance
    );
    const verificaOk = await verificaSaldoFinale(analisi.account.id);

    // Genera documentazione
    const doc = generaDoc(analisi, kontoEquity, registrazione, verificaOk);
    console.log('\n' + doc);

    // Salva documentazione
    const docFile = `chiusura-konto-1099-${new Date().toISOString().split('T')[0]}.txt`;
    fs.writeFileSync(docFile, doc);
    console.log(`📄 Documentazione salvata in: ${docFile}\n`);

    // Riepilogo
    console.log('='.repeat(60));
    console.log('  ✅ AUTOMAZIONE COMPLETATA');
    console.log('='.repeat(60));
    console.log(`\n✅ Registrazione: ${registrazione.move.name}`);
    console.log(`✅ Importo chiuso: CHF ${Math.abs(analisi.currentBalance).toFixed(2)}`);
    console.log(`${verificaOk ? '✅' : '⚠️ '} Saldo finale: ${verificaOk ? 'CHF 0.00' : 'Da verificare'}`);

    if (registrazione.move.state === 'draft') {
      console.log(`\n⚠️  IMPORTANTE: Vai su Odoo e valida la registrazione ${registrazione.move.name}`);
    }

    console.log('');

  } catch (error) {
    console.error('\n❌ ERRORE:', error.message);
    if (error.stack) console.error(error.stack);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = {
  connect,
  analizzaKonto1099,
  trovaContoPatrimonioNetto,
  creaRegistrazioneChiusura,
  verificaSaldoFinale
};
